### DISCUSSION TITLE
Start uvicorn server with reload programatically

### DISCUSSION THREAD

**Discussion by omerXfaruq** [Kludex/uvicorn#1482] [Ideas]
Hello!

Within the [Gradio](https://github.com/gradio-app/gradio/blob/9c1cef57f67694de29135fc6ea6f581572e7a324/gradio/networking.py#L156) project we are using uvicorn as our web server. However uvicorn does not support reload mode very dynamically, I am thinking of some hacks to use the reload mode. Could take your suggestions as well.

But I think it would be great for uvicorn to support reload mode programatically as well.
It currently requires it to be used like:

From `__main__` file
`uvicorn.run(app="main:app", reload=True")` 

and this would be great to have

From anywhere
`uvicorn.run(app=app_object, reload=True")`

**Comment by Kludex** [comment]
What's the difference? You can extract the path from it anyway. If we made the change, we'd just forward it as string anyway...

**Reply by omerXfaruq** [reply]
It's much more easier to use that way when the server is **in** a library or it is not just an web app.

For example I currently build it this way programatically

~~`gradio.launch(debug=True)`~~

**Reply by Kludex** [reply]
> It's much more easier to use that way when the server is an library or it is not just an web app.

Just to make it clear here, is this just because you don't want to extract the information from the app object? Don't get me wrong, that can a reason... I'm just not getting it how "much more easier" is we are talking about.

> For example I currently build it this way programatically
> 
> `gradio.launch(debug=True)`

Where's that? I can only see this: https://github.com/gradio-app/gradio/blob/9c1cef57f67694de29135fc6ea6f581572e7a324/gradio/networking.py#L38-L42 ...

**Reply by omerXfaruq** [reply]
My bad, should've shared an example, [here](https://github.com/gradio-app/gradio/blob/3fe954bd553531ea28ef7b9f77d3d4c509463bf0/demo/hello_world_3/run.py#L17) is one.

You can create ML Demos very easily using Gradio. And would like to add reloading support with `demo.launch(reload=True),` but it is hard to do when it is like below. Planning to hack it around but I think it would be useful to have it nonetheless.

```
From __main__ file
uvicorn.run(app="main:app", reload=True")

and this would be great to have

From anywhere
uvicorn.run(app=app_object, reload=True")
```

**Reply by omerXfaruq** [reply]
Btw unrelated point, I don't really like how we start uvicorn server in Gradio right now, planning to run it directly over uvicorn instead of running it in a thread. Could take your suggestions about it if you have any.

**Comment by omerXfaruq** [comment]
Found success like [this](https://github.com/gradio-app/gradio/pull/1278).
Though I cannot do it within the Gradio library and I think it is a limiting feature. If I could do it within the library I could just provide this kind of launch to users:

`demo.launch(reload=True)`


Btw is this limiting due to how uvicorn is designed or how python is designed? Because another [watchdog](https://pypi.org/project/watchdog/) library limits under `__name__ == "__main__" `as well:

**Reply by omerXfaruq** [reply]
@Kludex

**Reply by Kludex** [reply]
I was not getting what you mean... The quotes inside the app parameter really confused me.

Yes. It's a limitation. If you look for `__main__` [here](https://docs.python.org/3.8/library/multiprocessing.html#multiprocessing-programming), you can read more about it.

On my Mac, I don't have very useful logs... I remember I had an error in the past with my Ubuntu, I'll get it later on.
